heka/sandbox/decoders/extract_executive_summary.lua

```diff
@@ -0,0 +1,176 @@
+required to power the executive dashboard.
+
+See: https://bugzilla.mozilla.org/show_bug.cgi?id=1155871
+
+*Example Heka Configuration*
+
+.. code-block:: ini
+
+    [TelemetryInput]
+    # File or S3 input
+    decoder = "MultiDecoder"
+
+    [MultiDecoder]
+    subs = ["ProtobufDecoder", "ExecutiveSummary"] # use for S3 telemetry data
+    #subs = ["TelemetryDecoder", "ExecutiveSummary"] # use for S3 landfill data
+    cascade_strategy = "all"
+    log_sub_errors = true
+
+    [TelemetryDecoder]
+    type = "SandboxDecoder"
+    filename = "lua_decoders/extract_telemetry_dimensions.lua"
+    memory_limit = 100000000
+    output_limit = 1000000
+
+    [ExecutiveSummary]
+    type = "SandboxDecoder"
+    filename = "lua_decoders/extract_executive_summary.lua"
+    memory_limit = 100000000
+--]]
+
+require "cjson"
+local l = require "lpeg"
+require "string"
+
+local SEC_IN_HOUR = 60 * 60
+local SEC_IN_DAY = SEC_IN_HOUR * 24
+
+local normalize_channel =
+l.C"release" * -1 +
+l.C"beta" +
+(l.P("nightly") * -1 + "nightly-cck-") / "nightly" +
+l.C"aurora" * -1 +
+l.Cc"Other"
+
+local function anywhere (p)
+  return lpeg.P{ p + 1 * lpeg.V(1) }
+end
+
+local normalize_os =
+(l.P"Windows" + "WINNT") / "Windows" +
+l.P"Darwin" / "Mac" +
+(anywhere"Linux" + anywhere"BSD" + anywhere"SunOS") / "Linux" +
+l.Cc"Other"
+
+
+local function get_search_counts()
+    -- google, bing, yahoo, other
+    local cnts = {0, 0, 0, 0}
+    local json = read_message("Fields[payload.keyedHistograms]")
+    if not json then return cnts end
+
+    local ok, khist = pcall(cjson.decode, json)
+    if not ok then return cnts end
+    if type(khist.SEARCH_COUNTS) ~= "table" then return cnts end
+
+    for k, v in pairs(khist.SEARCH_COUNTS) do
+        for i, e in ipairs({"[Gg]oogle", "[Bb]ing", "[Yy]ahoo", "."}) do
+            if string.match(k, e) then
+                if type(v.sum) ~= "number" then return cnts end
+                cnts[i] = cnts[i] + v.sum
+                break
+            end
+        end
+    end
+    -- todo need the method for calculating the default searches on v4 data or
+    -- is it simply the count that matchers the SEARCH_DEFAULT_ENGINE entry?
+    return cnts;
+end
+
+
+local function get_hours()
+    local json = read_message("Fields[payload.info]")
+    local ok, json = pcall(cjson.decode, json)
+    if not ok then return 0 end
+    local uptime = json.subsessionLength
+
+    if type(uptime) ~= "number" or uptime < 0 or uptime >= 180 * SEC_IN_DAY then
+        return 0
+    end
+    uptime = uptime / SEC_IN_HOUR -- convert to hours
+    return uptime
+end
+
+
+local function get_default()
+    local json = read_message("Fields[environment.settings]")
+    local ok, json = pcall(cjson.decode, json)
+    if not ok then return false end
+
+    local default = json.isDefaultBrowser
+    if type(default) == "boolean" then
+        return default
+    end
+    return false
```

Comment:
It may be good to return true/false/nil to distinguish "unknown" from "not default"